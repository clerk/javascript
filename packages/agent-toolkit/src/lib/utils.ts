import type { ClerkTool } from './clerk-tool';
import type { ToolsContext } from './types';

// A helper type that maps T to a new type with every leaf replaced by R.
type DeepTransform<T, R> =
  T extends Array<infer U> ? DeepTransform<U, R>[] : T extends object ? { [K in keyof T]: DeepTransform<T[K], R> } : R;

/**
 * Recursively transforms every value in an object (or array) by applying a function.
 *
 * The result has the same structure as the input object,
  but each leaf value is replaced with the return type R of the transform function.
 */
export function deepTransform<T, R>(input: T, transformFn: (value: any) => R): DeepTransform<T, R> {
  if (Array.isArray(input)) {
    // Recursively transform each element of the array.
    return input.map(item => deepTransform(item, transformFn)) as DeepTransform<T, R>;
  } else if (input !== null && typeof input === 'object') {
    // Recursively transform each property of the object.
    const result: any = {};
    for (const key in input) {
      if (Object.prototype.hasOwnProperty.call(input, key)) {
        result[key] = deepTransform((input as any)[key], transformFn);
      }
    }
    return result as DeepTransform<T, R>;
  } else {
    // For non-objects, apply the transform function.
    return transformFn(input) as DeepTransform<T, R>;
  }
}

/**
 * A mapped type that replaces every property in T with type R.
 */
type ShallowTransform<T, R> = {
  [K in keyof T]: R;
};

/**
 * Transforms the top-level values of an object using a transformation function.
 *
 */
export function shallowTransform<T extends object, R>(
  input: T,
  transformFn: (value: T[keyof T]) => R,
): ShallowTransform<T, R> {
  const result = {} as ShallowTransform<T, R>;
  for (const key in input) {
    if (Object.prototype.hasOwnProperty.call(input, key)) {
      const typedKey = key as keyof T;
      result[typedKey] = transformFn(input[typedKey]);
    }
  }
  return result;
}

export const prunePrivateData = (context: ToolsContext, o?: Record<any, any> | null) => {
  if (context.allowPrivateMetadata) {
    return o;
  }

  if (o && o.private_metadata) {
    delete o.private_metadata;
  }
  return o;
};

/**
 * Filters tools based on a search pattern.
 * The pattern can be one of the following:
 * 1. The name of the category (e.g. "users") or the name of the category followed by .* (e.g. "users.*")
 * 2. The name of a specific tool within a category (e.g. "users.getCount")
 */
export const filterTools = (tools: Record<string, Record<string, ClerkTool>>, pattern: string): ClerkTool[] => {
  if (!pattern || pattern.length === 0) {
    throw new Error('No pattern specified');
  }

  if (pattern === '*') {
    return Object.values(tools).flatMap(category => Object.values(category));
  }

  const validPattern = /^[a-zA-Z0-9_]+(\.[a-zA-Z0-9_*]+)?$/;
  if (!validPattern.test(pattern)) {
    throw new Error('Invalid pattern');
  }

  const [category, tool] = pattern.split('.');
  if (!category || (category && !tools[category])) {
    throw new Error(`Tool category not found: ${category}`);
  }

  if ((category && tool === '*') || (category && !tool)) {
    return Object.values(tools[category]);
  }

  if (category && tool && !tools[category][tool]) {
    throw new Error(`Tool not found: ${tool}`);
  }

  return [tools[category][tool]];
};
