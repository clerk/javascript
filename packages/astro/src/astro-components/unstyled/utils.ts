/**
 * This function is used when an element is passed as a default slot and we need
 * to add an attribute to it so that we can reference it in a click listener.
 */
export function addUnstyledAttributeToFirstTag(html: string, attributeValue: string): string {
  return html.replace(/(<[^>]+)>/, `$1 data-clerk-unstyled-id="${attributeValue}">`);
}

/**
 * Logs a deprecation warning when the 'as' prop is used.
 */
export function logAsPropUsageDeprecation() {
  if (import.meta.env.PROD) {
    return;
  }

  console.warn(
    `[@clerk/astro] The 'as' prop is deprecated and will be removed in a future version. ` +
      `Use the default slot with the 'asChild' prop instead. `,
  );
}
