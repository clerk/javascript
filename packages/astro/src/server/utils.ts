import { constants } from '@clerk/backend/internal';

export function getAuthKeyFromRequest(req: Request, key: keyof typeof constants.Attributes): string | null | undefined {
  return getHeader(req, constants.Headers[key]);
}

function getHeader(req: Request, name: string) {
  return req.headers.get(name);
}

export const isRedirect = (res: Response) => {
  return (
    [300, 301, 302, 303, 304, 307, 308].includes(res.status) ||
    res.headers.get(constants.Headers.ClerkRedirectTo) === 'true'
  );
};

export const setHeader = <T extends Response>(res: T, name: string, val: string): T => {
  res.headers.set(name, val);
  return res;
};
