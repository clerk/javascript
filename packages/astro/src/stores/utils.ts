import type {
  ActiveSessionResource,
  InitialState,
  OrganizationCustomPermissionKey,
  OrganizationCustomRoleKey,
  OrganizationResource,
  Resources,
  UserResource,
} from '@clerk/types';

export function deriveState(clerkLoaded: boolean, state: Resources, initialState: InitialState | undefined) {
  if (!clerkLoaded && initialState) {
    return {
      ...deriveFromSsrInitialState(initialState),
      clerkLoaded,
    };
  }

  return {
    ...deriveFromClientSideState(state),
    clerkLoaded,
  };
}

function deriveFromSsrInitialState(initialState: InitialState) {
  const userId = initialState.userId;
  const user = initialState.user as any as UserResource;
  const sessionId = initialState.sessionId;
  const session = initialState.session as any as ActiveSessionResource;
  const organization = initialState.organization as any as OrganizationResource;
  const orgId = initialState.orgId;
  const orgRole = initialState.orgRole as OrganizationCustomRoleKey;
  const orgPermissions = initialState.orgPermissions as OrganizationCustomPermissionKey[];
  const orgSlug = initialState.orgSlug;
  const actor = initialState.actor;

  return {
    userId,
    user,
    sessionId,
    session,
    organization,
    orgId,
    orgRole,
    orgPermissions,
    orgSlug,
    actor,
  };
}

function deriveFromClientSideState(state: Resources) {
  const userId: string | null | undefined = state.user ? state.user.id : state.user;
  const user = state.user;
  const sessionId: string | null | undefined = state.session ? state.session.id : state.session;
  const session = state.session;
  const actor = session?.actor;
  const organization = state.organization;
  const orgId: string | null | undefined = state.organization ? state.organization.id : state.organization;
  const orgSlug = organization?.slug;
  const membership = organization
    ? user?.organizationMemberships?.find(om => om.organization.id === orgId)
    : organization;
  const orgPermissions = membership ? membership.permissions : membership;
  const orgRole = membership ? membership.role : membership;

  return {
    userId,
    user,
    sessionId,
    session,
    organization,
    orgId,
    orgRole,
    orgSlug,
    orgPermissions,
    actor,
  };
}
