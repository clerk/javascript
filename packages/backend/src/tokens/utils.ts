export function toSSRResource(resource?: { private_metadata: any } | { privateMetadata: any }) {
  // Delete sensitive private metadata from resource before rendering in SSR
  if (resource) {
    // @ts-expect-error
    delete resource['privateMetadata'];
    // @ts-expect-error
    delete resource['private_metadata'];
  }

  return resource;
}
