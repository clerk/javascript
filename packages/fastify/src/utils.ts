export const getSingleValueFromArrayHeader = (value?: string[] | string): string | undefined => {
  return (Array.isArray(value) ? value[0] : value) || undefined;
};
