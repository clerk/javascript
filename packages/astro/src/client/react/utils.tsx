import type { LoadedClerk } from '@clerk/types';
import { useStore } from '@nanostores/react';
import { computed } from 'nanostores';
import React from 'react';

import { $clerk, $csrState } from '../../stores/internal';

export const withClerk = <P extends { clerk: LoadedClerk | undefined | null }>(
  Component: React.ComponentType<P>,
  displayName?: string,
) => {
  displayName = displayName || Component.displayName || Component.name || 'Component';
  Component.displayName = displayName;
  const HOC = (props: Omit<P, 'clerk'>) => {
    const clerk = useStore(
      computed([$csrState, $clerk], (state, clerk) => {
        return state.isLoaded ? clerk : null;
      }),
    );

    return (
      <Component
        /**
         * Force the remount of the component if clerk is not loaded yet.
         * This is needed in order to avoid hydration errors in controlComponents.
         */
        key={clerk ? 'a' : 'b'}
        {...(props as P)}
        clerk={clerk}
      />
    );
  };
  HOC.displayName = `withClerk(${displayName})`;
  return HOC;
};

export type WithClerkProp<T = unknown> = T & {
  clerk: LoadedClerk | undefined | null;
};
