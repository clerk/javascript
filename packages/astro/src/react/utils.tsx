import type { LoadedClerk } from '@clerk/types';
import { computed, type Store, type StoreValue } from 'nanostores';
import React from 'react';

import { $clerk, $csrState } from '../stores/internal';

/**
 * This implementation of `useStore` is an alternative solution to the hook exported by nanostores
 * Reference: https://github.com/nanostores/react/blob/main/index.js
 */
function useStore<T extends Store, SV extends StoreValue<T>>(store: T): SV {
  const get = store.get.bind(store);
  return React.useSyncExternalStore(store.listen, get, get);
}

export const withClerk = <P extends { clerk: LoadedClerk | undefined | null }>(
  Component: React.ComponentType<P>,
  displayName?: string,
) => {
  displayName = displayName || Component.displayName || Component.name || 'Component';
  Component.displayName = displayName;
  const HOC = (props: Omit<P, 'clerk'>) => {
    const clerk = useStore(
      computed([$csrState, $clerk], (state, clerk) => {
        return state.isLoaded ? clerk : null;
      }),
    );

    return (
      <Component
        /**
         * Force the remount of the component if clerk is not loaded yet.
         * This is needed in order to avoid hydration errors in controlComponents.
         */
        key={clerk ? 'a' : 'b'}
        {...(props as P)}
        clerk={clerk}
      />
    );
  };
  HOC.displayName = `withClerk(${displayName})`;
  return HOC;
};

export type WithClerkProp<T = unknown> = T & {
  clerk: LoadedClerk | undefined | null;
};

// TODO-SHARED: Duplicate from @clerk/clerk-react
export const assertSingleChild =
  (children: React.ReactNode) =>
  (name: 'SignInButton' | 'SignUpButton' | 'SignOutButton' | 'SignInWithMetamaskButton') => {
    try {
      return React.Children.only(children);
    } catch (e) {
      return `You've passed multiple children components to <${name}/>. You can only pass a single child component or text.`;
    }
  };

// TODO-SHARED: Duplicate from @clerk/clerk-react
export const normalizeWithDefaultValue = (children: React.ReactNode | undefined, defaultText: string) => {
  if (!children) {
    children = defaultText;
  }
  if (typeof children === 'string') {
    children = <button type='button'>{children}</button>;
  }
  return children;
};

// TODO-SHARED: Duplicate from @clerk/clerk-react
export const safeExecute =
  (cb: unknown) =>
  (...args: any) => {
    if (cb && typeof cb === 'function') {
      return cb(...args);
    }
  };
