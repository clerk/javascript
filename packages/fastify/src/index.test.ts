jest.mock('./withClerkMiddleware', () => {
  return {
    withClerkMiddleware: () => 'withClerkMiddlewareMocked',
  };
});

import type { FastifyRequest } from 'fastify';

import * as api from './index';
import { createFastifyInstanceMock } from './test/utils';

describe('@clerk/fastify', () => {
  test('exports clerkPlugin, withAuth, getAuth', () => {
    expect(api).toMatchSnapshot();
  });

  describe('clerkPlugin()', () => {
    test('adds withClerkMiddleware as preHandler', () => {
      const doneFn = jest.fn();
      const fastify = createFastifyInstanceMock();

      api.clerkPlugin(fastify, {}, doneFn);

      expect(fastify.addHook).toBeCalledWith('preHandler', 'withClerkMiddlewareMocked');
      expect(doneFn).toBeCalled();
    });

    test('adds auth decorator', () => {
      const doneFn = jest.fn();
      const fastify = createFastifyInstanceMock();

      api.clerkPlugin(fastify, {}, doneFn);

      expect(fastify.decorateRequest).toBeCalledWith('auth', null);
      expect(doneFn).toBeCalled();
    });

    test('polyfills reply with server response methods', () => {
      const doneFn = jest.fn();
      const fastify = createFastifyInstanceMock();

      api.clerkPlugin(fastify, {}, doneFn);

      expect(fastify.decorateReply).toMatchSnapshot();
    });
  });

  describe('getAuth(req)', () => {
    test('returns req.auth', () => {
      const req = { key1: 'asa', auth: 'authObj' } as any as FastifyRequest;

      expect(api.getAuth(req)).toEqual('authObj');
    });

    test('returns {} when req.auth as default', () => {
      const req = { key1: 'asa' } as any as FastifyRequest;

      expect(api.getAuth(req)).toEqual({});
    });
  });
});
