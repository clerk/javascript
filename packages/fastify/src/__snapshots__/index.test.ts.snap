// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@clerk/fastify clerkPlugin() polyfills reply with server response methods 1`] = `
[MockFunction] {
  "calls": [
    [
      "setHeader",
      [Function],
    ],
    [
      "writeHead",
      [Function],
    ],
    [
      "end",
      [Function],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`@clerk/fastify exports clerkPlugin, getAuth 1`] = `
{
  "clerkClient": {
    "__unstable_options": {
      "apiUrl": "https://api.clerk.dev",
      "apiVersion": "v1",
      "jwtKey": "",
      "secretKey": "TEST_API_KEY",
    },
    "allowlistIdentifiers": de {
      "request": [Function],
    },
    "authenticateRequest": [Function],
    "clients": ce {
      "request": [Function],
    },
    "debugRequestState": [Function],
    "emailAddresses": me {
      "request": [Function],
    },
    "emails": ge {
      "request": [Function],
    },
    "interstitial": fe {
      "request": [Function],
    },
    "invitations": ye {
      "request": [Function],
    },
    "localInterstitial": [Function],
    "organizations": he {
      "request": [Function],
    },
    "phoneNumbers": Se {
      "request": [Function],
    },
    "redirectUrls": ke {
      "request": [Function],
    },
    "remotePrivateInterstitial": [Function],
    "remotePublicInterstitial": [Function],
    "remotePublicInterstitialUrl": [Function],
    "sessions": _e {
      "getSession": [Function],
      "getSessionList": [Function],
      "getToken": [Function],
      "request": [Function],
      "revokeSession": [Function],
      "verifySession": [Function],
    },
    "signInTokens": Ie {
      "request": [Function],
    },
    "smsMessages": Ae {
      "request": [Function],
    },
    "users": we {
      "request": [Function],
    },
  },
  "clerkPlugin": [Function],
  "createClerkClient": [Function],
  "getAuth": [Function],
}
`;

exports[`@clerk/fastify getAuth(req) throws error if clerkPlugin is on registered 1`] = `
"ðŸ”’ Clerk: The "clerkPlugin" should be registered before using the "getAuth".
Example:

import { clerkPlugin } from '@clerk/fastify';

const server: FastifyInstance = Fastify({ logger: true });
server.register(clerkPlugin);

For more info, check out the docs: https://clerk.dev/docs,
or come say hi in our discord server: https://rebrand.ly/clerk-discord
"
`;
